Adding option to install opensearch dashboards plugins

lieberlois opened the pull request:
This PR aims to close #378. I'm not sure what plugins to install, so for testing this, I need a list of plugins to install. 

Signed-off-by: Luis Schweigard <luis.schweigard@maibornwolff.de>

idanl21 commented:
Hey @lieberlois 
for testing the concept, you try and install the `alertingDashboards` for example.
We also need to see why it failed the TLS test

lieberlois commented on opensearch-operator/pkg/builders/cluster.go:
```diff
@@ -214,20 +214,7 @@ func NewSTSForNodePool(
-	} else {
-		mainCommand = []string{"/bin/bash", "-c", "./opensearch-docker-entrypoint.sh"}
-	}
+	mainCommand := helpers.BuildMainCommand("./bin/opensearch-plugin", cr.Spec.General.PluginsList, true, "./opensearch-docker-entrypoint.sh")
```
Ill answer in your other comment :)

lieberlois commented on opensearch-operator/pkg/helpers/reconcile-helpers.go:
```diff
@@ -138,3 +139,27 @@ func VersionCheck(instance *opsterv1.OpenSearchCluster) (int32, string) {
 	return httpPort, securityconfigPath
 }
+
+func BuildMainCommand(installerBinary string, pluginsList []string, batchMode bool, entrypoint string) []string {
```
Hi @idanl21 , yes I did 👍  wasn't sure if it was worth it, but there were some duplications, so I thought why not!

If you don't like the func I can revert the change of course!

lieberlois commented:
@idanl21 thanks for the recommendation, the test worked - however, alertingDashboards are preinstalled, which causes the opensearch dashboards plugin installer to fail.

This is part of the spec I used to test:
```yaml
apiVersion: opensearch.opster.io/v1
kind: OpenSearchCluster
metadata:
  name: opensearch-cluster
spec:
  general:
    serviceName: opensearch-cluster
    setVMMaxMapCount: true
    version: 2.4.1
    pluginsList: []
  dashboards:
    enable: true
    version: 2.4.1
    replicas: 1
    pluginsList:
      - alertingDashboards
```

These are the logs of the dashboards pod:
```sh
Attempting to transfer from alertingDashboards
Attempting to transfer from https://ci.opensearch.org/ci/dbc/distribution-build-opensearch-dashboards/2.4.1/latest/linux/arm64/tar/builds/opensearch-dashboards/plugins/alertingDashboards-2.4.1.zip
Transferring 7147209 bytes....................
Transfer complete
Retrieving metadata from plugin archive
Extracting plugin archive
Extraction complete
Plugin alertingDashboards already exists, please remove before installing a new version
```

As for the TLS tests: the tests are not very robust and fail occasionally. For many of my PRs and also local tests, just restarting the tests solved the issue 👍  in case it doesnt, we need to take a proper look into the TLS tests.

swoehrl-mw reviewed (changes requested):
Logic looks good. Missing for me:
* Unittest for the new functionality
* A paragraph in the userguide
* Adding the new field to the CRD reference

idanl21 commented:
> @idanl21 thanks for the recommendation, the test worked - however, alertingDashboards are preinstalled, which causes the opensearch dashboards plugin installer to fail.
> 
> This is part of the spec I used to test:
> 
> ```yaml
> apiVersion: opensearch.opster.io/v1
> kind: OpenSearchCluster
> metadata:
>   name: opensearch-cluster
> spec:
>   general:
>     serviceName: opensearch-cluster
>     setVMMaxMapCount: true
>     version: 2.4.1
>     pluginsList: []
>   dashboards:
>     enable: true
>     version: 2.4.1
>     replicas: 1
>     pluginsList:
>       - alertingDashboards
> ```
> 
> These are the logs of the dashboards pod:
> 
> ```shell
> Attempting to transfer from alertingDashboards
> Attempting to transfer from https://ci.opensearch.org/ci/dbc/distribution-build-opensearch-dashboards/2.4.1/latest/linux/arm64/tar/builds/opensearch-dashboards/plugins/alertingDashboards-2.4.1.zip
> Transferring 7147209 bytes....................
> Transfer complete
> Retrieving metadata from plugin archive
> Extracting plugin archive
> Extraction complete
> Plugin alertingDashboards already exists, please remove before installing a new version
> ```
> 
> As for the TLS tests: the tests are not very robust and fail occasionally. For many of my PRs and also local tests, just restarting the tests solved the issue 👍 in case it doesnt, we need to take a proper look into the TLS tests.

Hey  @lieberlois, so in that case, i believe that we need ti add the logic that prevent that error so client will not fail on it

lieberlois commented:
@swoehrl-mw I added 2 unit tests and the missing docs.

@idanl21 I don't think its necessary to catch this error - its the default behavior and consumers of this operator shouldn't install plugins that are already there; IMO it is the expected behavior for this to fail.

swoehrl-mw reviewed (changes requested):
Looks good. Only thing missing is adding the new field to the CRD reference (docs/designs/crd.md), then we can merge.

lieberlois commented:
@swoehrl-mw I just added the documentation for the `pluginsList` which by the way was also missing for the generalConfig aswell. I added it in both places :)
